File: src/main/java/com/boxfox/core/router/CoinRouter.java
package com.boxfox.core.router;

import com.boxfox.core.router.model.CoinPriceNetworkObject;
import com.boxfox.cross.common.data.PostgresConfig;
import com.boxfox.cross.service.CoinService;
import com.boxfox.vertx.router.*;
import com.boxfox.vertx.service.*;
import com.boxfox.cross.service.LocaleService;
import com.boxfox.cross.service.PriceService;
import com.google.api.client.http.HttpStatusCodes;
import com.linkbit.android.entity.CoinModel;
import io.one.sys.db.tables.daos.CoinDao;
import io.vertx.core.http.HttpMethod;
import io.vertx.ext.web.RoutingContext;

import java.util.ArrayList;
import java.util.List;

import static com.boxfox.cross.util.LogUtil.getLogger;

public class CoinRouter extends AbstractRouter {

    @Service
    private PriceService priceService;

    @Service
    private LocaleService localeService;

    @Service
    private CoinService coinService;

    @RouteRegistration(uri = "/coin/supported/list", method = HttpMethod.GET)
    public void getSupportWalletList(RoutingContext ctx) {
        getLogger().debug(String.format("Supported Coin Load : %s", ctx.request().remoteAddress().host()));
        this.coinService.getAllCoins(res -> {
            if (res.succeeded()) {
                ctx.response().end(gson.toJson(res.result()));
            } else {
                ctx.response().setStatusCode(HttpStatusCodes.STATUS_CODE_SERVER_ERROR).end();
            }
        });
    }

    @RouteRegistration(uri = "/coin/price", method = HttpMethod.GET)
    public void getCoinPrice(RoutingContext ctx, @Param(name = "symbol") String symbol, @Param(name = "locale") String locale) {
        localeService.getLocaleMoneySymbol(locale, res -> {
            if (res.succeeded()) {
                String unit = res.result();
                double price = priceService.getPrice(symbol, locale);
                if (price > 0) {
                    CoinPriceNetworkObject result = new CoinPriceNetworkObject();
                    result.setAmount(price);
                    result.setUnit(unit);
                    ctx.response().end(gson.toJson(result));
                } else {
                    ctx.response().setStatusCode(404).end();
                }
            } else {
                ctx.response().setStatusCode(500).end();
            }
        });
    }

}

File: src/main/java/com/boxfox/core/router/CoinRouter.kt
package com.boxfox.core.router

import com.boxfox.core.router.model.CoinPriceNetworkObject
import com.boxfox.linkbit.service.LocaleService
import com.boxfox.linkbit.service.coin.CoinService
import com.boxfox.linkbit.util.LogUtil.getLogger
import com.boxfox.vertx.router.AbstractRouter
import com.boxfox.vertx.router.Param
import com.boxfox.vertx.router.RouteRegistration
import com.boxfox.vertx.service.Service
import io.vertx.core.http.HttpMethod
import io.vertx.core.json.JsonArray
import io.vertx.ext.web.RoutingContext

class CoinRouter() : AbstractRouter() {

    @Service
    private lateinit var localeService: LocaleService

    @Service
    private lateinit var coinService: CoinService

    @RouteRegistration(uri = "/coins", method = [HttpMethod.GET])
    fun getSimpleCoinList(ctx: RoutingContext) {
        val locale = ctx.data()["locale"].toString()
        getLogger().debug(String.format("Coin Load : %s", ctx.request().remoteAddress().host()))
        this.coinService.getCoins(locale).subscribe({
            ctx.response().end(gson.toJson(it))
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/coins/price", method = [HttpMethod.GET])
    fun getCoinPriceList(ctx: RoutingContext) {
        val locale = ctx.data()["locale"].toString()
        getLogger().debug(String.format("Coin Load : %s", ctx.request().remoteAddress().host()))
        this.coinService.getPrices(locale).subscribe({
            ctx.response().end(gson.toJson(it))
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/coins/price", method = [HttpMethod.POST])
    fun getCoinPriceList(ctx: RoutingContext, @Param(name = "symbols") symbols: JsonArray) {
        val locale = ctx.data()["locale"].toString()
        getLogger().debug(String.format("Coin Load : %s", ctx.request().remoteAddress().host()))
        this.coinService.getPrices(symbols.map { it -> it.toString() }, locale).subscribe({
            ctx.response().end(gson.toJson(it))
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/coins/:symbol", method = [HttpMethod.GET])
    fun getCoinPrice(ctx: RoutingContext, @Param(name = "symbol") symbol: String) {
        val locale = ctx.data()["locale"].toString()
        localeService.getMoneySymbol(locale) { res ->
            if (res.succeeded()) {
                coinService.getPrice(symbol).subscribe({ price ->
                    val unit = res.result()
                    if (price > 0) {
                        val result = CoinPriceNetworkObject().apply {
                            this.symbol = symbol
                            this.amount = price
                            this.unit = unit
                        }
                        ctx.response().end(gson.toJson(result))
                    } else {
                        ctx.response().setStatusCode(404).end()
                    }
                },{
                    ctx.fail(it)
                })
            } else {
                ctx.fail(res.cause())
            }
        }
    }
}
